src/compute/src/sink/persist_sink.rs

```diff
@@ -130,111 +593,471 @@ where
+    let mut descriptions_input = append_op.new_input_connection(
+        batch_descriptions,
+        Exchange::new(move |_| hashed_id),
+        vec![Antichain::new()],
+    );
+    let mut batches_input = append_op.new_input_connection(
+        batches,
+        Exchange::new(move |_| hashed_id),
+        vec![Antichain::new()],
+    );
+
+    // Dropping this token signals that the operator should shut down cleanly.
+    let token = Rc::new(());
+    let token_weak = Rc::downgrade(&token);
+
+    // This operator accepts the current and desired update streams for a `persist` shard.
+    // It attempts to write out updates, starting from the current's upper frontier, that
+    // will cause the changes of desired to be committed to persist.
+
+    append_op.build_async(
+        scope,
+        move |mut capabilities, frontiers, scheduler| async move {
+            let mut cap_set = if active_worker {
+                CapabilitySet::from_elem(capabilities.pop().expect("missing capability"))
+            } else {
+                // Pop away, because we don't need it.
+                capabilities.pop().expect("missing capability");
+                CapabilitySet::new()
+            };
+
+            let mut description_buffer = Vec::new();
+            let mut batch_buffer = Vec::new();
+
+            // Contains descriptions of batches for which we know that we can
+            // write data. We got these from the "centralized" operator that
+            // determines batch descriptions for all writers.
+            let mut in_flight_descriptions: HashSet<(Antichain<Timestamp>, Antichain<Timestamp>)> =
+                HashSet::new();
+
+            let mut in_flight_batches: HashMap<
+                (Antichain<Timestamp>, Antichain<Timestamp>),
+                Vec<Batch<_, _, _, _>>,
+            > = HashMap::new();
+
+            // TODO(aljoscha): We need to figure out what to do with error results from these calls.
+            let persist_client = persist_clients
+                .lock()
+                .await
+                .open(persist_location)
+                .await
+                .expect("could not open persist client");
 
             let mut write = persist_client
                 .open_writer::<SourceData, (), Timestamp, Diff>(shard_id)
                 .await
                 .expect("could not open persist shard");
 
             while scheduler.notified().await {
-                let mut input_frontier = Antichain::new();
-                for frontier in frontiers.borrow().clone() {
-                    input_frontier.extend(frontier);
+                if token_weak.upgrade().is_none() {
+                    return;
                 }
 
-                if !active_write_worker
-                    || token_weak.upgrade().is_none()
-                    || shared_frontier.borrow().is_empty()
-                {
-                    return;
+                if !active_worker {
+                    continue;
```

Comment:
There was a bug in the async operator implementation that meant we have to stay alive, otherwise our peer operators would not see the frontier advance anymore. I don't know if this is still true but will double check.

Reply:
If it turns out this is still necessary to ensure correct behavior, a comment mentioning this would be great!